### ISSUE TITLE
Cannot get milestones or projects feature to work by title, only by number

### ISSUE THREAD

**Issue by JimBledsoe** [alstr/todo-to-issue-action#59]
I have created:
- public org project
- private org project
- public user project
- private user project
- issue level milestone in a private repo

I have created TODOs attempting to use all 5 cases above one at a time.  None of the issues get created.  Other TODOs with labels and assignees work fine.  The same tokens are used for all cases.  The token has pretty high access.

The milestone in a private repo may be an issue - have not duplicated the raw API call yet.

I used postman to hit the same API the action is using, giving the same accepts and authorization token headers and the public projects are returned.

Here is some markdown with the markdown file containing the TODOs and some of the debugging output from the processed run

---

# This is a markdown file to test the TODO generator

## Raw markdown source code

***The markdown below is written with an extra space in the begin comment so they show up here.  Remove the space between the opening <! to actually use these as valid comments in MarkDown.***

We can generate an issue with assignees and labels - this works fine

< !--- TODO A generated issue with assignee and labels
Please don't close this issue that was generated by automation.
assignees: JimBledsoe
labels: enhancement
-->

We can assign an issue to a milestone - not working

< !--- TODO A generated issue with a milestone in this repo
Please don't close this issue that was generated by automation.
milestone: Test Milestone
-->

We can assign an issue to a milestone in quotes perhaps? - not working and not in output below, as it was added later but same error

< !--- TODO A generated issue with a milestone in quotes in this repo
Please don't close this issue that was generated by automation.
milestone: "Test Milestone"
-->

We can assign an issue to a public user project - not working

< !--- TODO A generated issue with a public user project
user projects: JimBledsoe/Jim Public Project/To do
-->

We can assign an issue to a private user project - not working, but neither does GitHub API

< !--- TODO A generated issue with a private user project
user projects: JimBledsoe/Jim Private Project/To do
-->

We can assign an issue to a public org project - not working

< !--- TODO A generated issue with a public org project
org projects: stage3talent/S3T Public Project/To do
-->

We can assign an issue to a private org project - not working, but neither does GitHub API

< !--- TODO A generated issue with a private org project
org projects: stage3talent/S3T Private Project/To do
-->

## Errors from the Action Processing Stage

```bash
Run alstr/todo-to-issue-action@v4.0.6
/usr/bin/docker run --name a33c15488893555b5409196683167155c571a_30ec6f --label 8a33c1 --workdir /github/workspace --rm -e INPUT_TOKEN -e INPUT_PROJECTS_SECRET -e INPUT_CLOSE_ISSUES -e INPUT_AUTO_P -e INPUT_REPO -e INPUT_BEFORE -e INPUT_SHA -e INPUT_LABEL -e INPUT_COMMENT_MARKER -e INPUT_USER_PROJECTS -e INPUT_ORG_PROJECTS -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_RETENTION_DAYS -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e GITHUB_ACTION_REPOSITORY -e GITHUB_ACTION_REF -e GITHUB_PATH -e GITHUB_ENV -e RUNNER_OS -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/curriculum-GCP/curriculum-GCP":"/github/workspace" 8a33c1:5488893555b5409196683167155c571a
Processing issue 1 of 6
Issue created
Processing issue 2 of 6
Milestone Test Milestone does not exist! Dropping this parameter!
Issue created
Processing issue 3 of 6
Adding issue to user project 1 of 1
An error occurred, skipping
Issue created
Processing issue 4 of 6
Adding issue to user project 1 of 1
An error occurred, skipping
Issue created
Processing issue 5 of 6
Adding issue to org project 1 of 1
An error occurred, skipping
Issue created
Processing issue 6 of 6
Adding issue to org project 1 of 1
An error occurred, skipping
Issue created
```

## Workflow file

```yml
---
name: Create issues from TODO comments in source files
on:
  push:
#    branches:
#    - main
#    - master

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v2

      - name: Scan and Create issues for TODO comments
        uses: alstr/todo-to-issue-action@v4.0.6
        id: todo
        with:
          TOKEN: ${{ secrets.GITHUB_TOKEN }}
          PROJECTS_SECRET: ${{ secrets.GITHUB_TOKEN }}
          CLOSE_ISSUES: true
          AUTO_P: true
```

## Permissions of the Token

`admin:org, repo, workflow, write:packages`

**Comment by JimBledsoe** [comment]
Looking at the code and further testing with the API and the action, it looks like milestones will work if the number is specified instead of the title.  The documentation here shows using the title, not the number.

This comment worked, but not palatable:



API docs don't show how to get a milestone by anything other than the ID:
https://docs.github.com/en/rest/reference/issues#get-a-milestone

But there is an API that gets all the milestones for a repo to extract the number from:
https://docs.github.com/en/rest/reference/issues#list-milestones

**Comment by alstr** [comment]
Hi Jim, thanks for the detailed report.

From the logs you included it looks like there's an issue with the request to get the list of projects:

https://github.com/alstr/todo-to-issue-action/blob/70d10efd53a64c110938ec7d44ff36a045be76b4/main.py#L218

Or the list of columns:

https://github.com/alstr/todo-to-issue-action/blob/70d10efd53a64c110938ec7d44ff36a045be76b4/main.py#L234

If you do a Postman `GET` request to:

`'https://api.github.com/users/JimBledsoe/projects/'

Find the project ID for `JimPublicProject` and try it another `GET` request to:

`'https://api.github.com/projects/< project id >/columns/'

What response do you get?

As you identified, at the moment, milestones can only be specified by ID (number), as that is how the API works at a basic level. The ability to specify by title is an option for future, but it would involve an extra request somewhere. I haven't really had any feedback for the milestone implementation so not much idea on how people want it to work, but if this is preferred we should definitely look at it.

**Comment by JimBledsoe** [comment]
The columns_url returned from the project query for my user JimBledsoe private project, returned:
https://api.github.com/projects/12916614/columns

Trying that query in postman with the same 2 headers yields:
[
    {
        "url": "https://api.github.com/projects/columns/15218034",
        "project_url": "https://api.github.com/projects/12916614",
        "cards_url": "https://api.github.com/projects/columns/15218034/cards",
        "id": 15218034,
        "node_id": "PC_lALOAP-FvM4AxReGzgDoNXI",
        "name": "To do",
        "created_at": "2021-07-20T15:18:58Z",
        "updated_at": "2021-07-20T15:19:18Z"
    },
    {
        "url": "https://api.github.com/projects/columns/15218035",
        "project_url": "https://api.github.com/projects/12916614",
        "cards_url": "https://api.github.com/projects/columns/15218035/cards",
        "id": 15218035,
        "node_id": "PC_lALOAP-FvM4AxReGzgDoNXM",
        "name": "In progress",
        "created_at": "2021-07-20T15:18:58Z",
        "updated_at": "2021-07-20T15:18:58Z"
    },
    {
        "url": "https://api.github.com/projects/columns/15218036",
        "project_url": "https://api.github.com/projects/12916614",
        "cards_url": "https://api.github.com/projects/columns/15218036/cards",
        "id": 15218036,
        "node_id": "PC_lALOAP-FvM4AxReGzgDoNXQ",
        "name": "Done",
        "created_at": "2021-07-20T15:18:58Z",
        "updated_at": "2021-07-20T15:18:58Z"
    }
]

**Comment by JimBledsoe** [comment]
Yeah, both seem to be a case of your README saying its by name, but the implementation is by id for projects and number for milestones.

Either are achievable.  List of milestones to translate the names to numbers: https://api.github.com/repos/{org}/{repo}/milestones

**Comment by alstr** [comment]
I think we should separate milestone and project issues, yes. Regarding the project, I just added this to a test file and committed:

```python
    # TODO: Test this out
    #  user projects: alstr/Test Project/Test Col
```

When I go to the issue the project assignment is visible:

<img width="335" alt="Screenshot 2021-07-21 at 18 02 40" src="https://user-images.githubusercontent.com/11612438/126529605-1688c733-66cd-4f7a-9f07-858e787dd5d7.png">

Same if I go to the project:

<img width="715" alt="Screenshot 2021-07-21 at 17 59 36" src="https://user-images.githubusercontent.com/11612438/126529590-0f67914a-da11-4080-ab8f-5e06254ab3af.png">

I think your problem may be this in your workflow file:

`PROJECTS_SECRET: ${{ secrets.GITHUB_TOKEN }}`

As mentioned [here](https://github.com/alstr/todo-to-issue-action#projects), I recommend creating a new secret for projects:

> The action cannot access your projects by default. To enable access, you must create an encrypted secret in your repo settings, with the value set to a valid Personal Access Token. Then, assign the secret in the workflow file like PROJECTS_SECRET: ${{ secrets.PROJECTS_SECRET }}. Do not enter the raw secret.

The default provided `GITHUB_TOKEN` does not seem to grant access to projects from my experience. Just create a new token and give it `repo` and `workflow` access.

**Comment by JimBledsoe** [comment]
OMG, Yes, I had created a token for using this but did not specify it in the yml correctly.  I was still using the default secret and not the new org level one.  Well that explains why Postman was working and this was not.

Yes, I think everything in this thread is resolved.  I misread the milestone by number and had the wrong PAT being used.

Closing this as resolved.  I may put in an enhancement request for Milestone by Name at some point.  I need to see how the team responds to projects, but the other thread about project by repo is a sticky one for us.  I will comment on that thread.

Thanks for your time looking into this.
